src/Bounty.hs

```diff
@@ -184,56 +152,47 @@ correctCollection o c = case (destination c) of
 
 -- We need to make sure that the spending path is correct for the PotDatum TxOut TODO
 {-# INLINABLE validateUseOfPot #-}
-validateUseOfPot :: Bounty -> TxOut -> Maybe BountyDatum -> Maybe BountyDatum -> Bool
-validateUseOfPot bounty potTxOut mpot mcollection = case mpot of
-  Just m -> case mcollection of
-    Just (CollectionDatum c) ->
-      solidCollection bounty c
-        && correctCollection potTxOut c
-    _ -> False
-  _ -> False
+validateUseOfPot :: Bounty -> Maybe TxOut -> Maybe BountyDatum -> Maybe BountyDatum -> Bool
+-- validateUseOfPot bounty mPotTxOut mpot mcollection = case mpot of
+validateUseOfPot bounty (Just potTxOut) (Just _) (Just (CollectionDatum c)) =
+  solidCollection bounty c && correctCollection potTxOut c
+validateUseOfPot _ _ _ _ = False
 
 -- This needs to check that
 -- - Only valid voters are counted
 -- - All new voters have signed the tx.
 {-# INLINABLE validateKeyChanges #-}
 validateKeyChanges :: TxInfo -> [PubKeyHash] -> [PubKeyHash] -> [PubKeyHash] -> Bool
-validateKeyChanges info voters before after =
-  let newVotes = [a | a <- after, elem a voters]
-      compVal = [a | a <- after, elem a before]
-   in compVal == before
+validateKeyChanges info voters votesBefore votesAfter =
+  let newVotes = [a | a <- votesAfter, elem a voters]
+      compVal = [a | a <- votesAfter, elem a votesBefore]
+   in compVal == votesBefore
         && all (txSignedBy info) newVotes
 
 -- High-Level Functions -- ehh lmao
 {-# INLINABLE containsClass #-}
 containsClass :: TxOut -> AssetClass -> Bool
 containsClass o a = (assetClassValueOf (txOutValue o) a) > 0
 
-{-# INLINABLE getOutput #-}
-getOutput :: [TxOut] -> AssetClass -> TxOut
-getOutput txOuts asset =
-  case [o | o <- txOuts, containsClass o asset] of
-    [x] -> x
-    _   -> traceError "Fail here."
+{-# INLINABLE findOutputForClass #-}
+findOutputForClass :: AssetClass -> [TxOut] -> Maybe TxOut
+findOutputForClass asset = find $ \o -> containsClass o asset
 
 {-# INLINABLE containsPot #-}
 containsPot :: TxInfo -> TxOut -> Bool
 containsPot info o =
-  let d = potDatum info o
-   in case d of
-        Just PotDatum -> True
-        _             -> False
+  case findBountyDatum info o of
```

Comment:
Since Haskell is expression based, you don't need to assign value to a variable and then pattern match on that value, you can directly pattern match on the value/expression.